File: GreenGate v2/frontend/src/api/csvUpload.ts
import client from './client';

export interface CsvUploadEntry {
  scope: number;
  category: string;
  activity_data: number;
  activity_unit: string;
  emission_factor: number;
  co2e_tonnes: number;
  transaction_count: number;
}

export interface UncategorisedRow {
  row_number: number;
  description: string;
  amount: number;
  category: string;
}

export interface CsvUploadResult {
  entries_created: number;
  categorised_count: number;
  uncategorised_count: number;
  total_co2e_tonnes: number;
  entries: CsvUploadEntry[];
  uncategorised: UncategorisedRow[];
}

export interface FileResult {
  filename: string;
  entries_created: number;
  co2e_tonnes: number;
  errors: string | null;
}

export interface FolderScanResult {
  files_processed: number;
  total_entries_created: number;
  total_co2e_tonnes: number;
  file_results: FileResult[];
}

export const uploadCsv = async (
  file: File,
  reportingYear: number = 2025,
): Promise<CsvUploadResult> => {
  const formData = new FormData();
  formData.append('file', file);
  const res = await client.post<CsvUploadResult>(
    `/emissions/upload/csv?reporting_year=${reportingYear}`,
    formData,
    { headers: { 'Content-Type': 'multipart/form-data' } },
  );
  return res.data;
};

export const scanFolder = async (
  folderPath: string,
  reportingYear: number = 2025,
): Promise<FolderScanResult> => {
  const res = await client.post<FolderScanResult>('/emissions/upload/folder', {
    folder_path: folderPath,
    reporting_year: reportingYear,
  });
  return res.data;
};

File: GreenGate v2/backend/app/schemas/csv_upload.py
from __future__ import annotations

from decimal import Decimal
from typing import List, Optional

from pydantic import BaseModel


class CsvUploadEntry(BaseModel):
    scope: int
    category: str
    activity_data: float
    activity_unit: str
    emission_factor: float
    co2e_tonnes: float
    transaction_count: int


class UncategorisedRow(BaseModel):
    row_number: int
    description: Optional[str] = None
    amount: Optional[float] = None
    category: Optional[str] = None


class CsvUploadResponse(BaseModel):
    entries_created: int
    categorised_count: int
    uncategorised_count: int
    total_co2e_tonnes: float
    entries: List[CsvUploadEntry]
    uncategorised: List[UncategorisedRow]


class FolderScanRequest(BaseModel):
    folder_path: str
    reporting_year: int = 2025


class FileResult(BaseModel):
    filename: str
    entries_created: int
    co2e_tonnes: float
    errors: Optional[str] = None


class FolderScanResponse(BaseModel):
    files_processed: int
    total_entries_created: int
    total_co2e_tonnes: float
    file_results: List[FileResult]
